### ISSUE TITLE
Introduce `GraphqlSchemaNode` 

### ISSUE THREAD

**Issue by andimarek** [graphql-java/graphql-java#1251]
Currently we misuse `GraphQLType` as a general `GraphQLSchemaNode`. We should introduce an extra Interface above `GraphQLType` so that for example `GraphQLFieldDefinition` and `GraphQLNonNull` is not a `GraphQLType` but just a `GraphQLSchemaNode`.

It needs to be explored how big this refactoring is really is.

**Comment by bbakerman** [comment]
I think the name "Node" is wrong.  Just doesnt sit nice.  Node invokes the AST nodes to me.

All the "types" have GraphqlXXX so I would continue that.

Naming is hard but

* GraphqlSchemaItem
* GraphqlItem
* GraphqlComponent
* GraphqlElement
* GraphqlUnit

**Comment by andimarek** [comment]
GraphqlSchemaElement maybe?

**Comment by bbakerman** [comment]
The method

```
    TraversalControl accept(TraverserContext<GraphQLType> context, GraphQLTypeVisitor visitor);
```
is problematic in naming terms.  That is we could change it to be 

```
    TraversalControl accept(TraverserContext<? extends GraphQlSchemaElement > context, GraphQLTypeVisitor visitor);
```

but the visitor is named `GraphQLTypeVisitor` and that can be changed without making a major breaking change to that SPI

So we want to make that change??

**Comment by andimarek** [comment]
as inspiration: graphql-js has the following type:

```
export type GraphQLNamedType =
  | GraphQLScalarType
  | GraphQLObjectType
  | GraphQLInterfaceType
  | GraphQLUnionType
  | GraphQLEnumType
  | GraphQLInputObjectType;
```
to distinguish between types with names and wrapper types (non null, list)

**Comment by andimarek** [comment]
this will be resolved by #1605
